src/Databases/DatabaseFilesystem.cpp

```diff
@@ -0,0 +1,221 @@
+    extern const int FILE_DOESNT_EXIST;
+}
+
+DatabaseFilesystem::DatabaseFilesystem(const String & name_, const String & path_, ContextPtr context_)
+    : IDatabase(name_), WithContext(context_->getGlobalContext()), path(path_), log(&Poco::Logger::get("DatabaseFileSystem(" + name_ + ")"))
+{
+    fs::path user_files_path;
+    const auto & application_type = context_->getApplicationType();
+
+    if (application_type != Context::ApplicationType::LOCAL)
+        user_files_path = fs::canonical(fs::path(getContext()->getUserFilesPath()));
+
+    if (fs::path(path).is_relative())
+        path = user_files_path / path;
+    else if (application_type != Context::ApplicationType::LOCAL && !pathStartsWith(fs::path(path), user_files_path))
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Path must be inside user-files path ({})", user_files_path.string());
+
+    path = fs::absolute(path).lexically_normal().string();
+
+    if (!fs::exists(path))
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Path does not exist ({})", path);
+}
+
+std::string DatabaseFilesystem::getTablePath(const std::string & table_name) const
+{
+    fs::path table_path = fs::path(path) / table_name;
+    return table_path.lexically_normal().string();
+}
+
+void DatabaseFilesystem::addTable(const std::string & table_name, StoragePtr table_storage) const
+{
+    std::lock_guard lock(mutex);
+    auto [_, inserted] = loaded_tables.emplace(table_name, table_storage);
+    if (!inserted)
+        throw Exception(
+            ErrorCodes::LOGICAL_ERROR,
+            "Table with name `{}` already exists in database `{}` (engine {})",
+            table_name,
+            getDatabaseName(),
+            getEngineName());
+}
+
+bool DatabaseFilesystem::checkTableFilePath(const std::string & table_path, ContextPtr context_, bool throw_on_error) const
+{
+    // If run in Local mode, no need for path checking.
+    bool need_check_path = context_->getApplicationType() != Context::ApplicationType::LOCAL;
+    std::string user_files_path = fs::canonical(fs::path(context_->getUserFilesPath())).string();
+
+    // Check access for file before checking its existence
+    if (need_check_path && !fileOrSymlinkPathStartsWith(table_path, user_files_path))
+    {
+        if (throw_on_error)
+            throw Exception(ErrorCodes::DATABASE_ACCESS_DENIED, "File is not inside {}", user_files_path);
+        else
+            return false;
+    }
+
+    // Check if the corresponding file exists
+    if (!fs::exists(table_path) || !fs::is_regular_file(table_path))
+    {
+        if (throw_on_error)
+            throw Exception(ErrorCodes::FILE_DOESNT_EXIST, "File does not exist ({})", table_path);
+        else
+            return false;
+    }
+
+    return true;
+}
+
+bool DatabaseFilesystem::isTableExist(const String & name, ContextPtr context_) const
+{
+    {
+        std::lock_guard lock(mutex);
+        if (loaded_tables.find(name) != loaded_tables.end())
+            return true;
```

Comment:
Fixed, added cache invalidation

IMHO, it doesn't make much sense, because cache is only used to avoid doing all checks and validations every time. If a file is deleted or moved, the FILE_DOESNT_EXIST exception from StorageFile will be thrown when you try to read it